Implement markov(history=1) by scanning discrete latent variables

fehiepsi opened the pull request:
Changes to support experiments in https://github.com/pyro-ppl/numpyro/issues/678.

### Tasks
- [x] Support enum in scan
- [x] Add some tests
- [x] Add some examples

eb8680 commented on numpyro/contrib/control_flow/scan.py:
```diff
@@ -194,7 +270,14 @@ def g(*args, **kwargs):
+            with LocalNamedMessenger():
+                dim_to_name = msg["infer"].get("dim_to_name")
+                to_funsor(msg["value"], dim_to_name=OrderedDict([(k, dim_to_name[k]) for k in sorted(dim_to_name)]))
+                apply_stack(msg)
```
This logic is fine, the `LocalNamedMessenger` here ensures that these dimensions have no effect on the rest of model execution.

eb8680 commented on numpyro/contrib/funsor/enum_messenger.py:
```diff
@@ -253,9 +253,9 @@ def _get_dim_to_name(batch_shape, dim_to_name=None, dim_type=DimType.LOCAL):
-                continue
             name = dim_to_name.get(dim, None)
+            if batch_shape[dim] == 1 and name is None:
+                continue
```
Per discussion on Slack, this change was made because the time dimension on the left sometimes necessitates empty dimensions appearing before they have been assigned a name.

fehiepsi replied:
yeah, I'll add this comment. thanks! :D

eb8680 commented on numpyro/contrib/funsor/enum_messenger.py:
```diff
@@ -501,6 +501,8 @@ def process_message(self, msg):
                 msg.get("done", False) or msg["is_observed"] or msg["infer"].get("expand", False) or \
                 msg["infer"].get("enumerate") != "parallel" or (not msg["fn"].has_enumerate_support):
+            if msg["type"] == "control_flow":
+                msg["kwargs"]["enum"] = True
```
nit: this should probably live in `infer_config`

fehiepsi replied:
Hmm, I need `scan_wrapper` knows that `enum=True` but `scan_wrapper` only accesses to `args`, `kwargs`, not `infer` field.

eb8680 replied:
Sorry, I meant that this logic setting `msg["kwargs"]["enum"] = True` should live in `infer_config`, not in `enum`.

fehiepsi replied:
Do you suggest that when we use `config_enumerate(fn)`, we will add `enum` kwargs to `control_flow` msg? Currently, `infer_config` does
```
    if msg["type"] in ("sample", "param"):
            msg["infer"].update(self.config_fn(msg))
```
so we need to change it to
```
        if msg["type"] in ("sample", "param"):
            msg["infer"].update(self.config_fn(msg))
        if msg["type"] == "control_flow":
         msg["kwargs"].update(self.config_fn(msg))
```
and in `config_enumerate`, we need to check if `msg` is `control_flow` to return `{"enum": True}`. It is a bit verbose to me and violate the purpose of `infer_config`. In addition, we don't need to use `config_enumerate` if users already provide `infer={"enumerate": "parallel"}` in `sample` statements. In that case, the only handler that affects `control_flow` is `enum`.

eb8680 replied:
> In that case, the only handler that affects control_flow is enum.

Ah, yes, good point, I guess what you have here is fine.

fehiepsi commented on numpyro/contrib/funsor/infer_util.py:
```diff
@@ -70,6 +71,28 @@ def config_fn(site):
+            prev_to_curr[k] = curr_var
+        prev_and_curr = prev_vars | frozenset(curr_vars)
+        with funsor.interpreter.interpretation(funsor.terms.lazy):
+            eliminate = (sum_vars | prod_vars) - prev_and_curr
```
Time variables do not appear in `sum_vars` and `prod_vars` so I think this is fine.

eb8680 replied:
If you have a time-global variable that couples a bunch of time-local variables (like a global parameter in a state-space model's transition matrix), then eliminating that global variable couples all of the local variables and breaks the conditional independence that allows the use of `sequential_sum_product`.  I think you'll need to run `sequential_sum_product` before this step, rather than after.

fehiepsi replied:
Hmm, I think I understand this wrong. Using `eliminate=frozenset()` passes `plate` and global discrete tests.

eb8680 replied:
Yes, when `eliminate == frozenset()` this works fine, but when that's not the case it may be incorrect to sum out variables in this order.

fehiepsi replied:
It seems that the test for multiple plates at the same dimension does not pass due to this issue. I think I need to follow your strategy: using `sequential_sum_product` before this step. Did you suggested that we need to eliminate time dimensions in `x[t]|x[t-1]` and `y[t]|y[t-1]` separately?

fehiepsi replied:
Yes, so I think we need to set eliminate=frozenset(). The tricky thing is when we have two plate statements at dim=-1, z[t]|x[t] and y[t]|z[t], we need to concatenate z and y, eliminate dimensions, then split the result. The current code will allocate two separate dimensions for two plates, which leads to wrong result. Do you have a solution for this?

fehiepsi replied:
Because we don't eliminate any variable, I think we can use `sum(log_factors)` here for simplicity. However, we still need to have a mechanism to merge variables that allocate the same dimension. It is the job of `sequential_sum_product` I guess?

eb8680 replied:
> However, we still need to have a mechanism to merge variables that allocate the same dimension

Sorry, I'm not sure I understand the question.  What do you mean here? What is the difference between `z[t, i]` and `z[t]`, which I assume are supposed to be `.inputs` of a factor funsor?  In this function, all of the log-factors are funsors, so I don't see why you can't just sum the log-factors.

fehiepsi replied:
Yes, we can take the sum, but assume z has addition plate D1 at dim=-1 and y has addition plate D2 at dim=-1, taking the sum will return a funsor with inputs time x prev x curr x D1 x D2. But I think the correct behavior is time x prev x curr x (D1 + D2).

eb8680 replied:
> but assume z has addition plate D1 at dim=-1 and y has addition plate D2 at dim=-1

Sorry, I still don't understand what you mean when you say that a funsor input `z` has an additional plate at dim=-1. Inputs are just inputs; all plates should be included in the log-factor funsors as additional inputs.  Can you give a concrete example set of log-factor funsors that exhibit this behavior?

eb8680 replied:
I see, I understand the situation you have in mind in `test_scan_enum_plates_same_dim`. I don't see why the fact that the plates share a dimension (dim=-1) should matter from the perspective of funsor, since everything is named and unordered.  However, I do understand now what you mean about stacking the plate dimensions (`D1 + D2`) rather than broadcasting to the cartesian product (`D1 x D2`).  

You can run tensor variable elimination (`funsor.sum_product.sum_product`) to eliminate those plates before applying `sequential_sum_product`, as you were doing before.  This would have the same effect as stacking the plates, but there are subtle restrictions on what can happen inside of them in the model for variable elimination to remain tractable - conservatively, any enumeration inside should be restricted to a single timestep (so the value of an enumerated sample site within one of the plates may not leave it or be used at the next timestep).  The situation in `test_scan_enum_plates_same_dim` is a special case of this where no new variables are enumerated within either plate.

fehiepsi replied:
Thanks a lot, @eb8680!!! I just eliminate all dimensions that do not appear in markov sites (discrete sites that depend on the last step) and ... all test cases that I can think about passed. Could you take a look at the tests and suggest to me other scenarios that we need to tests?

fehiepsi replied:
Looking like I am ready to port `hmm_enum` example now. :D thanks for your help!

fehiepsi commented on numpyro/contrib/funsor/infer_util.py:
```diff
@@ -70,6 +71,28 @@ def config_fn(site):
     return infer_config(fn, config_fn)
 
 
+def compute_markov_factors(time_to_factors, markov_vars, sum_vars, prod_vars):
```
Assume `x` is a discrete latent, `markov_vars[time]` is [`_init/x`]. If there are two discrete latents, `markov_vars[time]` is [`_init/x`, `_init/w`]. `sum_vars` are all variables `_init/x`, `_init_w`, `x`, `w`, `y`. `prod_vars` contains plate dimensions (inside and outside scan).

fehiepsi replied:
I just remove `sum_vars` and rename `markov_vars` to `init_vars` for clarity.

fehiepsi replied:
> the expected .inputs and .output of the return value and the arguments in time_to_factors

`time_to_factors[time]` contains all log_probs with time `time`. I think all of them have `reals()` outputs. `inputs` are `dim_to_name` at those sites having `time` dimension.

fehiepsi commented on numpyro/contrib/funsor/infer_util.py:
```diff
@@ -109,11 +134,26 @@ def model(*args, **kwargs):
 
+    if len(time_to_factors) > 0:
+        markov_factors = compute_markov_factors(time_to_factors, markov_vars, sum_vars, prod_vars)
+        log_factors = log_factors + markov_factors
```
Good question! `plate` inside and outside `scan` should be good, but using `scan` inside `plate` does not work (we can easily move `plate` inside `scan` though). However, I think the scan `transition` might not work if it depends on an outside discrete latent variable (I am not sure, I need to add a test if it works).

> no factors outside of the scan can depend on fresh variables from inside the scan

I am not sure. Factors outside of the scan can depend on the last carry value. I'll need to add a test for it.

fehiepsi replied:
For factors depend on the last x, I guess we need to reshape the last carry value... I am not sure if dimensions are aligned correctly. Let me test this first

fehiepsi replied:
To me: make sure to add test for plate.

fehiepsi replied:
@eb8680 It seems that `markov` does not allow outside-dependency
```python
import jax.numpy as jnp
import funsor
import numpyro
from numpyro.contrib.control_flow import scan
from numpyro.contrib.funsor import config_enumerate, markov, to_funsor, trace, enum, log_density
import numpyro.distributions as dist

data = jnp.repeat(jnp.array([0, 1]), (7, 4))
transition_probs = jnp.array([[0.8, 0.2], [0.1, 0.9]])
locs = jnp.array([-1.0, 1.0])

def model(data):
    x = None
    for i, y in markov(enumerate(data)):
        probs = jnp.array([0.5, 0.5]) if x is None else transition_probs[x]
        x = numpyro.sample(f"x_{i}", dist.Categorical(probs))
        numpyro.sample(f"y_{i}", dist.Normal(locs[x], 1), obs=y)
    numpyro.sample("w", dist.Normal(jnp.zeros(2)[x]))

with trace() as tr, enum():
    numpyro.handlers.seed(config_enumerate(model), rng_seed=1)(data)
print(tr["w"]["infer"])
log_density(enum(config_enumerate(model)), (data,), {}, {'w': jnp.ones(2)})[0]
```
returns
```
{'dim_to_name': OrderedDict([(-1, '_pyro_dim_1')])}
DeviceArray([-15.72769, -15.72769], dtype=float32)
```
Is this a bug or unsupported?

eb8680 replied:
> Is this a bug or unsupported?

This is unsupported under `markov` as it is currently defined, but it's a common enough scenario that we should probably change `markov` to support it.

fehiepsi replied:
I just made a change that guarantees shape of the last carry is consistent w.r.t. sequential strategy. For now, do you think that we can add a note that there should be no variable depends on the last carry? (or there is a simple fix here?)

eb8680 replied:
You can just add a note about the restriction for now.  I'm not sure there's a quick fix, although I do think it should be changed at some point.

fehiepsi replied:
Sure, I'll add to the docs

fehiepsi commented on test/contrib/test_funsor.py:
```diff
@@ -188,3 +193,94 @@ def testing():
+
+    actual_log_joint = log_density(enum(config_enumerate(fun_model)), (data,), {}, {})[0]
+    expected_log_joint = log_density(enum(config_enumerate(fun_model)), (data,), {}, {})[0]
+    assert_allclose(actual_log_joint, expected_log_joint)
```
Ops, I tested fun_model to itself.........

fehiepsi commented on numpyro/contrib/funsor/infer_util.py:
```diff
@@ -109,11 +134,26 @@ def model(*args, **kwargs):
+
+            time_dim = None
+            for dim, name in dim_to_name.items():
+                if name.startswith("_time"):
```
I explained the name in the docs of `scan`.

fehiepsi commented on test/contrib/test_funsor.py:
```diff
@@ -188,3 +193,198 @@ def testing():
+    assert_allclose(actual_log_joint, expected_log_joint)
+
+
+def test_scan_enum_plates_same_dim():
```
@eb8680 I meant this test.

fehiepsi commented:
@eb8680 @neerajprad Looking like all tests and examples pass. Compiling time is much more reasonable now (without truncation, model 1 took less than 1m in my slow Windows subsystem for Linux). :D 

@neerajprad Could you help me store the data file to cloudfront?

fehiepsi commented on examples/hmm_enum.py:
```diff
@@ -0,0 +1,279 @@
+                    numpyro.sample("y", dist.Bernoulli(probs_y[x.squeeze(-1)]), obs=y)
+        return (x, t + 1), None
+
+    x_init = jnp.zeros((num_sequences, 1), dtype=jnp.int32)
```
Instead of `x_init = 0` in Pyro, we need to make `x_init` have the same shape as the output (in non-enumerate version).

fehiepsi commented on examples/hmm_enum.py:
```diff
@@ -0,0 +1,279 @@
+density in parallel. In addition, the stacked form allows us
+to use the parallel-scan algorithm in [2], which reduces parallel
+complexity from O(length) to O(log(length)).
+
```
@eb8680 Could you suggest me a paragraph mentioning about how to obtain the data and how to preprocess it?

eb8680 replied:
I think the data you're using is a single batch from the music dataset used in the Pyro HMM and DMM examples. The relevant Pyro file, with a bit of commentary, is [here](https://github.com/pyro-ppl/pyro/blob/dev/examples/dmm/polyphonic_data_loader.py).  Unfortunately the preprocessing logic there depends heavily on PyTorch, so the easiest thing to do is probably to do the preprocessing once in Pyro, save the resulting transformed data, and store it in Cloudfront.

fehiepsi replied:
I see. I think we just need to mention what is the structure of the data. E.g. in `pyro`, the data has 88 tones but it seems to only have 44 tones here.

eb8680 replied:
I think there are only 44 tones here because the batch I grabbed only happened to contain 44 notes.  I would suggest using the full dataset instead now that the compilation issues have been resolved.

fehiepsi replied:
I can confirm that there are 44 notes with `num_sequences=17`. :D

eb8680 commented on examples/hmm_enum.py:
```diff
@@ -0,0 +1,279 @@
+
+    def transition_fn(carry, y):
+        x_prev, t = carry
+        with numpyro.plate("sequences", num_sequences, dim=-2):
```
This is a nit for these models, but these `sequences` plates should really live outside of `transition_fn` to avoid tempting users to potentially break enumeration by enumerating variables inside plates across timesteps.  Were there issues supporting that? You may need to do the same trick with `plate` that you did with `trace` and `substitute`, where you push the `plate` handler inside of the `scan` wrapper, but that doesn't seem so bad.

fehiepsi replied:
Do you mean to support something like
```
with plate(...):
    scan(...)
```
? If so, I think that it is tricky to put time dimensions at the right position (it seems to be doable but might require non-enum users write correct plate codes). Could you give an example on how things could go wrong so I can add to the docs instead?

eb8680 replied:
All of the [restrictions from Pyro's enumeration tutorial](http://pyro.ai/examples/enumeration.html#Dependencies-among-plates) still apply.  A simple failure case would be having an enumerated variable outside of a plate depend on a variable enumerated inside that plate at the previous timestep (`x_t ~ p( . | x_t-1, y_i_t-1)`, `y_t_i ~ p( . | x_t)`).

fehiepsi replied:
Thanks, I didn't know about those restrictions. I'll add a link to it and learn more about them.

eb8680 commented on numpyro/contrib/control_flow/scan.py:
```diff
@@ -78,7 +83,99 @@ def _subs_wrapper(subs_map, i, length, site):
+    # XXX: This implementation only works for history size=1 but can be
+    # extended to history size > 1 by running `f` `history_size` times
+    # for initialization. However, `sequential_sum_product` does not
+    # support history size > 1, so we skip supporting it here.
```
Note `funsor.sum_product.sarkka_bilmes_product` does support `history > 1`.

fehiepsi replied:
Oh, I didn't know about that. Nice, thanks for pointing out! I temp into trying it this weekend. :D

eb8680 commented on numpyro/contrib/control_flow/scan.py:
```diff
@@ -78,7 +83,99 @@ def _subs_wrapper(subs_map, i, length, site):
+            nonlocal carry_shape_at_t1
+            carry_shape_at_t1 = [jnp.shape(x) for x in tree_flatten(new_carry)[0]]
+            # make new_carry have the same shape as carry
+            # FIXME: is this rigorous?
```
You may want to add some unit tests for this

fehiepsi replied:
The thing is it passes all the test and I can't figure out when it possibly fails. :(

eb8680 replied:
Things may get more complicated if you decide to support `history > 1` and there are enumerated variables with different periods.

eb8680 commented on numpyro/contrib/funsor/infer_util.py:
```diff
@@ -70,6 +71,33 @@ def config_fn(site):
+        # remove `_init/` prefix to convert prev to curr
+        prev_to_curr = {k: "/".join(k.split("/")[1:]) for k in prev_vars}
+        # we eliminate all plate dimensions not available at discrete sites.
+        eliminate_vars = (sum_vars | prod_vars) - time_to_markov_dims[time_var]
```
Does this `eliminate_vars` include variables only enumerated *within* a single timestep in `scan`? For example, suppose that for whatever reason you had a discrete HMM whose emission distributions were Gaussian mixtures rather than Gaussian - you can eliminate the latent mixture component in the emission distribution in the initial `funsor.sum_product.sum_product` call below, but not the latent state of the markov chain.

fehiepsi replied:
Does this [test](https://github.com/pyro-ppl/numpyro/pull/682/files#diff-fed0396c70de3697c3e0eb518df45b00R303) cover that usage case? Here we not eliminate only dimensions in the latent state of the markov chain.

eb8680 replied:
Yes - if you run that test and set a debugger breakpoint here, what are the contents of `eliminate_vars`?  Does it contain `w`?

fehiepsi replied:
Yes, it is `frozenset({'y', '_init/w', '_init/y', 'D', 'w'}`. Is it expected?

eb8680 replied:
Yeah, `w`, `_init/w` and `D` are expected.  `y` shouldn't be in `sum_vars` in the first place, since it's not an input of any of the log-factor funsors. I left a comment indicating where that should be fixed.

eb8680 commented on numpyro/contrib/control_flow/scan.py:
```diff
@@ -78,7 +83,99 @@ def _subs_wrapper(subs_map, i, length, site):
+                trace = {}
+        else:
+            with handlers.block(), packed_trace() as trace, promote_shapes(), enum(), markov():
+                new_carry, y = config_enumerate(seeded_fn)(carry, x)
```
Add a comment explaining what's happening here for maintainability.

eb8680 commented on numpyro/contrib/control_flow/scan.py:
```diff
@@ -78,7 +83,99 @@ def _subs_wrapper(subs_map, i, length, site):
 
-def scan_wrapper(f, init, xs, length, reverse, rng_key=None, substitute_stack=[]):
+class promote_shapes(Messenger):
+    # a helper messenger to promote shapes of `fn` and `value`
```
nit: add a comment or docstring with example old and new shapes for `fn` and `value` for maintainability.

fehiepsi commented:
Thanks for reviewing Eli, I will address your comments soon! I am not sure how large the "batch" of the original dataset is, probably too large to run MCMC? About nested scan, I haven't tested if it works out of box because I didn't design `scan` to be used in nested pattern. Let me test and if it doesn't work, I'll make an issue for it. For `plates` outside `scan`, I think that we can create a separate issue. The most complicated issue is to move "time" dimension to the correct position for data of observed sites in `transition_fn`, of condition/substitute handlers, and all the collected `sample` msg of latent sites. It might be nice to have but I am not sure if there is a simple way to do it.

eb8680 commented:
> I am not sure how large the "batch" of the original dataset is

I wouldn't be surprised if NumPyro's HMC + `sequential_sum_product` could handle the smallest of the four original music datasets directly, or could at least do so with the help of the parallel scan primitive from #686 and a GPU or TPU.  If it does end up being too large, I suggest switching to another dataset or loading a full music dataset and taking a slice here for reproducibility, since I don't remember which one I took these from...

> For plates outside scan, I think that we can create a separate issue.

It's up to you to decide what to include in the upcoming release, but since support for this is a notable difference from enumeration in Pyro, I suggest adding some logic in this PR to raise a `NotImplementedError` in `enum.process_msg` if a plate handler is on the handler stack, with a helpful error message that would explain how the user could fix their model.  Ditto for nested `scan` - have `enum` raise a `NotImplementedError` with an informative error message unless/until you support that case.  You may also want to add xfailing tests for both situations.

I'll have to think a bit more about whether there's an easier implementation strategy for supporting plates outside `scan`.

fehiepsi commented:
Thanks Eli! Currently, we raised `RuntimeError` for `scan` under `plate`. I have switched it to `NotImplementedError` and told users to see `scan` docs for example on how to move `plate` into `scan`. About nested scan, it does not work even for continuous case due to substitute logic, which does not . I have added a sentence there tell users that nested scan is not supported in the docs. I'll study a bit about the data to see if we can use the original version here.

fehiepsi commented:
@eb8680 Thanks for throughout reviews! I think I have addressed most of your comments. The remaining one is at `eliminate_vars` which I am not sure if the current behavior is at your expectation. If you notice that I missed something else, please let me know.

fehiepsi commented:
I also want to comment that although compiling time is not an issue anymore, MCMC is still slow for this dataset, especially at early steps where NUTS takes long trajectory to get u-turn/divergence. Probably it is faster in GPU because we have vectorized computations but I don't have GPU to test (my desktop is unusable after delivered to my new home :( ).

eb8680 commented on numpyro/contrib/funsor/infer_util.py:
```diff
@@ -109,11 +140,35 @@ def model(*args, **kwargs):
+            if time_dim is None:
+                log_factors.append(log_prob)
+
             sum_vars |= frozenset({site['name']})
```
Don't add `site['name']` to `sum_vars` here if it was observed

fehiepsi replied:
Thanks, I just made the commit.

eb8680 commented:
> Probably it is faster in GPU because we have vectorized computations but I don't have GPU to test

@fehiepsi I ran `python hmm_enum.py -m 2 --device gpu --num-sequences 110` on a cloud GPU (Tesla K80, ~12GB RAM) and it finished in a couple hours. I suspect that GPU could just about handle the entire dataset (`--num-sequences 229`), but there seems to be some initial processing (which I haven't looked into) that uses quite a bit more GPU memory than the actual sampling steps and causes an out-of-memory allocation error when I try to increase `num_sequences` further.

fehiepsi commented:
@eb8680 Before sampling, we try to find valid initial parameters, which involves evaluating the joint density. With hidden_dim=16, tones=51, num_sequences=229, length=129, the log_prob at `x` has shape `128 x 16 x 16 x 229 x 1` and at `y` is `128 x 16 x 1 x 229 x 51`, which are a bit large for involved operators I guess.

eb8680 commented:
> Before sampling, we try to find valid initial parameters, which involves evaluating the joint density

Hmm, would it be possible to add initialization strategies that don't enumerate discrete variables?  That seems like a useful optimization, although probably not worth looking at before the release.

fehiepsi commented:
We find initial parameters sequentially, so memory requirement should be similar to sampling. I think you can verify it by separating out `compile` and `sample` steps as in [this benchmark](https://github.com/pyro-ppl/numpyro/blob/benchmarks/benchmarks/hmm.py#L91). I think your observation of high memory usage before the actual sampling might not be relevant to it.

fehiepsi commented:
@eb8680 I tried to inspect the execution in different ways and observed the same thing as yours: memory used for initialization is much higher than sampling...

fehiepsi commented:
I think the issue is `numpyro.plate` returns `None` instead of the list of indices. Could you verify if it fixes the issue in your system? It fixes the issue in my system.

eb8680 commented:
Nice! Now the HMM examples run on the full dataset on that same GPU.
